diff --git a/onnxruntime/test/contrib_ops/skiplayernorm_op_test.cc b/onnxruntime/test/contrib_ops/skiplayernorm_op_test.cc
index 4e4a2eee3f..551558c7f4 100644
--- a/onnxruntime/test/contrib_ops/skiplayernorm_op_test.cc
+++ b/onnxruntime/test/contrib_ops/skiplayernorm_op_test.cc
@@ -7,6 +7,12 @@
 #include "test/common/cuda_op_test_utils.h"
 #include "test/providers/provider_test_utils.h"
 
+/*#if defined(USE_DNNL)
+#define DNNL_GTEST_SKIP(message) GTEST_SKIP_(message)
+#else
+#define DNNL_GTEST_SKIP(message)
+#endif*/
+
 namespace onnxruntime {
 namespace test {
 constexpr float epsilon_ = 1e-12f;
@@ -61,7 +67,7 @@ static void RunTest(
 
   auto rocm_ep = DefaultRocmExecutionProvider();
   auto dml_ep = DefaultDmlExecutionProvider();
-  auto dnnl_ep = DefaultDnnlExecutionProvider();
+  //auto dnnl_ep = DefaultDnnlExecutionProvider();
   if (!use_float16) {
     OpTester test(op_type.c_str(), 1, onnxruntime::kMSDomain);
     test.AddInput<float>("input", input_dims, input_data);
@@ -759,6 +765,8 @@ TEST(SkipLayerNormTest, SkipLayerNormBatch2_Skip_Broadcast_No_Batch_Size) {
       0.46111652255058289, -0.21333980560302734, -0.29631003737449646, 3.5148544311523438,
       0.28433859348297119, -0.17090578377246857, -0.92897164821624756, 4.6924152374267578,
       0.46111652255058289, -0.21333980560302734, -0.29631003737449646, 3.5148544311523438};
+  // Skip this test case if DNNL is available but disabled---Broadcasting is not available for DNNL yet
+  //DNNL_GTEST_SKIP("Skipping due to disabled DNNL");
 
   RunTest(input_data,
           skip_data,
@@ -806,6 +814,8 @@ TEST(SkipLayerNormTest, SkipLayerNormBatch2_Skip_Broadcast_Batch_Size_1) {
       0.46111652255058289, -0.21333980560302734, -0.29631003737449646, 3.5148544311523438,
       0.28433859348297119, -0.17090578377246857, -0.92897164821624756, 4.6924152374267578,
       0.46111652255058289, -0.21333980560302734, -0.29631003737449646, 3.5148544311523438};
+  // Skip this test case if DNNL is available but disabled---Broadcasting is not available for DNNL yet
+  //DNNL_GTEST_SKIP("Skipping due to disabled DNNL");
 
   RunTest(input_data,
           skip_data,
