--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -120,2 +120,8 @@
+option(MATSDK_DISABLE_LOGGING "Compile internal SDK logging out" OFF)
+if(MATSDK_DISABLE_LOGGING)
+  target_compile_definitions(matsdk_internal_config INTERFACE
+    MATSDK_DISABLE_LOGGING)
+endif()
+
 option(GCC5_CXX11_ABI_WORKAROUND "Workaround: Use legacy C++11 ABI (for GCC 5 compatibility)" OFF)
 if(GCC5_CXX11_ABI_WORKAROUND)

--- a/lib/include/mat/config-default.h
+++ b/lib/include/mat/config-default.h
@@ -29,7 +29,9 @@
 #endif
 #define HAVE_MAT_JSONHPP
 #define HAVE_MAT_ZLIB
+#if !defined(MATSDK_DISABLE_LOGGING)
 #define HAVE_MAT_LOGGING
+#endif
 /* #define HAVE_MAT_WIN_LOG     */
 /* #define HAVE_MAT_EVT_TRACEID     */
 #define HAVE_MAT_STORAGE

--- a/lib/pal/PAL.cpp
+++ b/lib/pal/PAL.cpp
@@ -183,3 +183,4 @@
 #if !defined(_WIN32) && defined(__linux__)
+#ifdef HAVE_MAT_LOGGING
         static std::mutex m;
         static std::map<std::thread::id, pid_t> threads;
@@ -194,4 +195,5 @@
         }
 #else
 #define     gettid()       std::this_thread::get_id()
+#endif
 #endif

--- a/lib/pal/TaskDispatcher_CAPI.cpp
+++ b/lib/pal/TaskDispatcher_CAPI.cpp
@@ -44,4 +44,5 @@
                 }
                 catch (const std::exception& ex) {
+                    (void)ex;
                     LOG_ERROR("Unhandled exception in CAPI task: %s", ex.what());
                 }

--- a/lib/pal/WorkerThread.cpp
+++ b/lib/pal/WorkerThread.cpp
@@ -247,4 +247,5 @@
                         }
                         catch (const std::exception& ex) {
+                            (void)ex;
                             LOG_ERROR("Unhandled exception in worker task: %s", ex.what());
                         }

--- a/lib/pal/posix/sysinfo_utils_mac.mm
+++ b/lib/pal/posix/sysinfo_utils_mac.mm
@@ -8,4 +8,5 @@
 #include <unistd.h>
 #include <uuid/uuid.h>
 #include <vector>
+#include <TargetConditionals.h>
 #import <Foundation/Foundation.h>
@@ -37,5 +38,8 @@
 std::string GetDeviceId()
 {
+#if !TARGET_OS_OSX
+    return {EMPTY_GUID};
+#else
     uuid_t uuidBytes;
     const struct timespec spec = {1, 0};
     int hostUUIDResult = gethostuuid(uuidBytes, &spec);
@@ -49,2 +53,3 @@
     return {EMPTY_GUID};
+#endif
 }
