diff --git a/onnxruntime/test/providers/checkers.cc b/onnxruntime/test/providers/checkers.cc
index f1a7240ea3..436031dfa8 100644
--- a/onnxruntime/test/providers/checkers.cc
+++ b/onnxruntime/test/providers/checkers.cc
@@ -154,6 +154,7 @@ struct TensorCheck<int8_t> {
     }
 
     const bool has_abs_err = params.absolute_error.has_value();
+    const int8_t default_abs_err = 1;
     if (has_abs_err) {
       double threshold = *(params.absolute_error);
 
@@ -162,7 +163,8 @@ struct TensorCheck<int8_t> {
       }
     } else {
       for (int i = 0; i < size; ++i) {
-        EXPECT_EQ(cur_expected[i], cur_actual[i]) << "i:" << i;
+        // EXPECT_EQ(cur_expected[i], cur_actual[i]) << "i:" << i;
+        EXPECT_NEAR(cur_expected[i], cur_actual[i], default_abs_err) << "i:" << i;
       }
     }
   }
