diff --git a/src/dawn/native/ShaderModule.h b/src/dawn/native/ShaderModule.h
index e474ea4588..54aea404da 100644
--- a/src/dawn/native/ShaderModule.h
+++ b/src/dawn/native/ShaderModule.h
@@ -205,7 +205,7 @@ enum class TextureQueryType : uint8_t { TextureNumLevels, TextureNumSamples };
     X(uint32_t, group)                   \
     X(uint32_t, binding)
 DAWN_SERIALIZABLE(struct, TextureMetadataQuery, TEXTURE_METADATE_QUERY_MEMBER) {
-    using TextureQueryType = TextureQueryType;
+    using TextureQueryType = detail::TextureQueryType;
 };
 #undef TEXTURE_METADATE_QUERY_MEMBER
 
