diff --git a/src/tint/api/BUILD.cmake b/src/tint/api/BUILD.cmake
index 6372c4ee77..2f98c30e13 100644
--- a/src/tint/api/BUILD.cmake
+++ b/src/tint/api/BUILD.cmake
@@ -58,6 +58,7 @@ tint_target_add_dependencies(tint_api lib
   tint_lang_wgsl_common
   tint_lang_wgsl_features
   tint_lang_wgsl_inspector
+  tint_lang_wgsl_inspector
   tint_lang_wgsl_program
   tint_lang_wgsl_sem
   tint_lang_wgsl_writer_ir_to_program
