diff --git a/src/ngraph/CMakeLists.txt b/src/ngraph/CMakeLists.txt
index d733ff0e3..98fe09253 100644
--- a/src/ngraph/CMakeLists.txt
+++ b/src/ngraph/CMakeLists.txt
@@ -708,6 +708,14 @@ if(NGRAPH_LIB_VERSIONING_ENABLE)
         VERSION ${NGRAPH_VERSION}
         SOVERSION ${NGRAPH_API_VERSION})
 endif()
+
+if(WIN32)
+    set_target_properties(ngraph PROPERTIES RUNTIME_OUTPUT_NAME "ovep_ngraph")
+    set_target_properties(ngraph PROPERTIES ARCHIVE_OUTPUT_NAME "ovep_ngraph")
+else()
+    set_target_properties(ngraph PROPERTIES LIBRARY_OUTPUT_NAME "ovep_ngraph")
+endif()
+
 if(NGRAPH_JSON_ENABLE)
     target_link_libraries(ngraph PRIVATE libjson)
 endif()
