--- absl/container/internal/layout.h	2023-11-28 09:35:48
+++ absl/container/internal/layout.updated.h	2023-11-28 10:13:14
@@ -181,9 +181,11 @@
 #include <sanitizer/asan_interface.h>
 #endif
 
-#if defined(__GXX_RTTI)
-#define ABSL_INTERNAL_HAS_CXA_DEMANGLE
-#endif
+// Comment out ABSL_INTERNAL_HAS_CXA_DEMANGLE definition to work around this issue:
+// https://github.com/abseil/abseil-cpp/issues/1435
+// #if defined(__GXX_RTTI)
+// #define ABSL_INTERNAL_HAS_CXA_DEMANGLE
+// #endif
 
 #ifdef ABSL_INTERNAL_HAS_CXA_DEMANGLE
 #include <cxxabi.h>
