diff --git a/cmake/platform.cmake b/cmake/platform.cmake
index fa51aa7..3d24fdc 100644
--- a/cmake/platform.cmake
+++ b/cmake/platform.cmake
@@ -64,9 +64,6 @@ elseif(UNIX OR APPLE OR MINGW)
         # unconditionnaly.
         set(CMAKE_CCXX_FLAGS "${CMAKE_CCXX_FLAGS} -Wno-pass-failed")
     elseif("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
-        if(NOT CMAKE_CXX_COMPILER_VERSION VERSION_LESS 5.0)
-            set(DEF_ARCH_OPT_FLAGS "-march=native -mtune=native")
-        endif()
         if(CMAKE_CXX_COMPILER_VERSION VERSION_LESS 6.0)
             # suppress warning on assumptions made regarding overflow (#146)
             set(CMAKE_CCXX_FLAGS "${CMAKE_CCXX_FLAGS} -Wno-strict-overflow")
