diff --git a/CMakeLists.txt b/CMakeLists.txt
index 8ab0422..c18d70f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -11,6 +11,16 @@ include(CMakePackageConfigHelpers)
 include(CTest)
 include(GNUInstallDirs)
 
+function(onnxruntime_add_include_to_target dst_target)
+    foreach(src_target ${ARGN})
+      if(TARGET ${src_target})
+        target_include_directories(${dst_target} PRIVATE $<TARGET_PROPERTY:${src_target},INTERFACE_INCLUDE_DIRECTORIES>)
+        target_compile_definitions(${dst_target} PRIVATE $<TARGET_PROPERTY:${src_target},INTERFACE_COMPILE_DEFINITIONS>)
+        target_sources(${dst_target} PRIVATE $<TARGET_PROPERTY:${src_target},INTERFACE_SOURCES>)
+      endif()
+    endforeach()
+endfunction()
+
 option(BUILD_SHARED_LIBS "build shared libraries" OFF)
 option(RE2_USE_ICU "build against ICU for full Unicode properties support" OFF)
 
@@ -156,7 +166,7 @@ endif()
 
 foreach(dep ${ABSL_DEPS})
   string(REGEX REPLACE "^absl_" "absl::" dep ${dep})
-  target_link_libraries(re2 PUBLIC ${dep})
+  onnxruntime_add_include_to_target(re2 ${dep})
 endforeach()
 
 if(RE2_USE_ICU)
