--- pytorch/cmake/Modules/FindNCCL.cmake	2020-05-15 01:03:07.938426870 +0000
+++ FindNCCL.cmake.fix_compile	2020-07-04 15:58:05.328284969 +0000
@@ -73,6 +73,7 @@
 ")
     try_run(NCCL_VERSION_MATCHED compile_result ${PROJECT_BINARY_DIR} ${file}
           RUN_OUTPUT_VARIABLE NCCL_VERSION_FROM_HEADER
+          CMAKE_FLAGS  "-DINCLUDE_DIRECTORIES=${NCCL_INCLUDE_DIRS}"
           LINK_LIBRARIES ${NCCL_LIBRARIES})
     if (NOT NCCL_VERSION_MATCHED)
       message(FATAL_ERROR "Found NCCL header version and library version do not match! \
