--- a/include/cutlass/arch/arch.h
+++ b/include/cutlass/arch/arch.h
@@ -79,43 +79,43 @@

 ////////////////////////////////////////////////////////////////////////////////////////////////////
 struct Sm50 {
-  static int const kMinComputeCapability = 50;
+  static inline constexpr int kMinComputeCapability = 50;
 };
 struct Sm60 {
-  static int const kMinComputeCapability = 60;
+  static inline constexpr int kMinComputeCapability = 60;
 };
 struct Sm61 {
-  static int const kMinComputeCapability = 61;
+  static inline constexpr int kMinComputeCapability = 61;
 };
 struct Sm70 {
-  static int const kMinComputeCapability = 70;
+  static inline constexpr int kMinComputeCapability = 70;
 };
 struct Sm72 {
-  static int const kMinComputeCapability = 72;
+  static inline constexpr int kMinComputeCapability = 72;
 };
 struct Sm75 {
-  static int const kMinComputeCapability = 75;
+  static inline constexpr int kMinComputeCapability = 75;
 };
 struct Sm80 {
-  static int const kMinComputeCapability = 80;
+  static inline constexpr int kMinComputeCapability = 80;
 };
 struct Sm86 {
-  static int const kMinComputeCapability = 86;
+  static inline constexpr int kMinComputeCapability = 86;
 };
 struct Sm89 {
-  static int const kMinComputeCapability = 89;
+  static inline constexpr int kMinComputeCapability = 89;
 };
 struct Sm90 {
-  static int const kMinComputeCapability = 90;
+  static inline constexpr int kMinComputeCapability = 90;
 };


 struct Sm100 {
-  static int const kMinComputeCapability = 100;
+  static inline constexpr int kMinComputeCapability = 100;
 };

 struct Sm101 {
-  static int const kMinComputeCapability = 101;
+  static inline constexpr int kMinComputeCapability = 101;
 };

 struct Sm120 {
-  static int const kMinComputeCapability = 120;
+  static inline constexpr int kMinComputeCapability = 120;
 };

 /// Triggers a breakpoint on the device
