--- a/lib/system/EventProperties.cpp
+++ b/lib/system/EventProperties.cpp
@@ -449,7 +449,7 @@
     evt_prop* EventProperties::pack()
     {
         size_t size = m_storage->properties.size() + m_storage->propertiesPartB.size() + 1;
-        evt_prop * result = static_cast<evt_prop *>(calloc(sizeof(evt_prop), size));
+        evt_prop * result = static_cast<evt_prop *>(calloc(size, sizeof(evt_prop)));
         if (result==nullptr)
         {
             LOG_ERROR("Unable to allocate memory to pack EventProperties");
